examples/ode-integrator.dx

```diff
@@ -1,22 +1,41 @@
+'## Euler Integrator
+Using only generic vector operations.
 
-EState : Type -> Type = \d. d=>Real
-Time :Type = Real
-Dynamics : Type -> Type = \d. EState d -> EState d
+Time : Type = Real
 
+def eulerStep (dict: VSpace a) ?=>
+              (zref: Ref h a) ->
+              (dynamics: a -> Time -> a) ->
+              (dt: Time) ->
+              (t: Time)
+              : {State h} Unit =
+    z = get zref
+    zref := z + dt .* dynamics z t
 
-eulerStep : Ref h (EState d) -> Dynamics d -> Time -> {State h} Unit =
-  \sRef dyn dt.
-    s = get sRef
-    sRef := s + dt .* dyn s
+def euler (dict: VSpace a) ?=> 
+          (dynamics: a -> Time -> a) ->
+          (z0: a) ->
+          (t0: Time) ->
+          (t1: Time) ->
+          (dt: Time)
+          : a =
+     n = floor ((t1 - t0) / dt)
+     snd $ withState z0 \zref.
+       for i:(Fin n).
+         t = t0 + (i2r (ordinal i)) * dt
+         eulerStep zref dynamics dt t
 
-euler : Int -> Dynamics d -> EState d -> Time -> EState d =
-  \n f x0 dt.
-    snd $ withState x0 \sRef.
-      for i:(Fin n). eulerStep sRef f dt
+def myDyn (dict: VSpace a) ?=> 
+          (z : a) -> (t:Time) : a = z
 
-myDyn : Dynamics d = \d. d
+z0 = [1.0]
+t0 = 0.0
+t1 = 1.0
+dt = 0.0001
 
-s0 = [1.0]
-
-:p euler 100000 myDyn s0 0.00001
+:p euler myDyn z0 t0 t1 dt
 > [2.7182682]
+
+:plot
+  xsTest = linspace (Fin 100) 0.0 1.0
+  zip xsTest $ map (\t1. sum $ euler myDyn z0 t0 t1 dt) xsTest
```

Comment:
Just to clarify my understanding: currently this is repeating the integration between `t0` and `t1` at each `t1` value in `xsTest` right? (As opposed to integrating from `t0` to `t_1`, then `t_1` to `t_2`, etc.)

I'm still getting familiar with the `dex` syntax and prelude, but I'm wondering (not necessarily for now, or part of this commit, of course!) if there's an easier solution with [`scan`](https://github.com/google-research/dex-lang/blob/main/prelude.dx#L240)? It may be able to avoid explicitly dealing with `State` at all, which could be nice for the step function. I'm not sure without exploring a bit though.

This is a pretty cool example. 😄

Reply:
Hmm... seems the CI build is getting a very slightly different result for the example integration, and wants a:
```
> <graphical output>
```
at the end.

Reply:
Yeah I think you're right. It would be better to use a scan here!

However, note that scan/fold and the `State` effect are very similar, and in fact the [`scan`/`fold` implementations in prelude](https://github.com/google-research/dex-lang/blob/main/prelude.dx#L240-L247) make use of the `State` effect! Basically, `scan` and `fold` are two ways to perform some iteration with state, but they're both generalized by `withState`, which also allows state can change during an iteration, including through functions that take the reference as an argument.

Reply:
To answer @lancelet 's question:  Yes, the plotting is repeating the integration.  I'm going to try this morning to let the integrator take in a vector of output times, and output the state at those times.  This should be a nice demo of dependent types, since those two lists will have to have the same length.